crates/rune-modules/src/time.rs

```diff
@@ -29,53 +29,659 @@
+    pub fn from_millis(millis: u64) -> Self {
+        Self {
+            inner: tokio::time::Duration::from_millis(millis),
+        }
+    }
+
+    /// Creates a new `Duration` from the specified number of microseconds.
+    ///
+    /// # Examples
+    ///
+    /// ```rune,no_run
+    /// use time::Duration;
+    ///
+    /// let duration = Duration::from_micros(1_000_002);
+    /// ```
+    #[rune::function(keep, path = Self::from_micros)]
+    pub fn from_micros(micros: u64) -> Self {
+        Self {
+            inner: tokio::time::Duration::from_micros(micros),
+        }
+    }
+
+    /// Creates a new `Duration` from the specified number of nanoseconds.
+    ///
+    /// Note: Using this on the return value of `as_nanos()` might cause unexpected behavior:
+    /// `as_nanos()` returns a u128, and can return values that do not fit in u64, e.g. 585 years.
+    /// Instead, consider using the pattern `Duration::new(d.as_secs(), d.subsec_nanos())`
+    /// if you cannot copy/clone the Duration directly.
+    ///
+    /// # Examples
+    ///
+    /// ```rune,no_run
+    /// use time::Duration;
+    ///
+    /// let duration = Duration::from_nanos(1_000_000_123);
+    /// ```
+    #[rune::function(keep, path = Self::from_nanos)]
+    pub fn from_nanos(nanos: u64) -> Self {
+        Self {
+            inner: tokio::time::Duration::from_nanos(nanos),
+        }
+    }
+
+    /// Returns the number of seconds contained by this `Duration` as `f64`.
+    ///
+    /// The returned value does include the fractional (nanosecond) part of the duration.
+    ///
+    /// # Examples
+    ///
+    /// ```rune,no_run
+    /// use time::Duration;
+    ///
+    /// let duration = Duration::from_secs(60).as_secs_f64();
+    /// ```
+    #[rune::function(keep, path = Self::as_secs_f64)]
+    pub fn as_secs_f64(&self) -> f64 {
+        self.inner.as_secs_f64()
+    }
+
+    /// Creates a new `Duration` from the specified number of seconds represented
+    /// as `f64`.
+    ///
+    /// # Examples
+    ///
+    /// ```rune,no_run
+    /// use time::Duration;
+    ///
+    /// let duration = Duration::from_secs_f64(0.0);
+    /// ```
+    #[rune::function(keep, path = Self::from_secs_f64)]
+    pub fn from_secs_f64(secs: f64) -> VmResult<Self> {
+        match tokio::time::Duration::try_from_secs_f64(secs) {
+            Ok(duration) => VmResult::Ok(Self { inner: duration }),
+            Err(e) => vm_panic!(e),
+        }
+    }
 }
 
-/// Sleep for the given [`Duration`].
-/// 
-/// # Examples
-/// 
-/// ```rune,no_run
-/// use time::Duration;
+impl Duration {
+    /// Converts [`Duration`] into a [`std::time::Duration`].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use time::Duration;
+    ///
+    /// let duration = Duration::from_secs(5);
+    /// let std_duration = duration.into_std();
+    /// ```
+    pub fn into_std(self) -> std::time::Duration {
+        std::time::Duration::new(self.inner.as_secs(), self.inner.subsec_nanos())
+    }
```

Comment:
Considering this, we could make all methods private and eliminate the `keep` parameters (`rune::function`), handling the conversion between Duration's externally instead.

Reply:
I'm fine either way. The `rune-modules` crate is currently unstable, so deciding what should and should not be public is up for future debate.

I have a minor preference to err on the side of caution just allowing for conversion seems like a good approach since you don't expect the rune-specific methods to be easily usable.

Reply:
Sure, that's fine with me as well!